lib/transfer-memory-allocator.js

```diff
@@ -0,0 +1,542 @@
+                        const availableSpace = nextNodeStartIndex - (currentNodeEndIndex + 1);
+                        if (bufferSize <= availableSpace) {
+                            const bufferBlockStartIndex = currentNodeEndIndex + 1; /* smallest possible next available index (lowest memory start value) */
+                            const allocatedBufferBlock = this[PRIVATE].pooledBuffer.subarray(
+                                bufferBlockStartIndex,
+                                bufferBlockStartIndex + bufferSize
+                            );
+
+                            foundBufferBlock = new TransferMemoryBlock(bufferSize, bufferBlockStartIndex, allocatedBufferBlock);
+
+                            // add node into linked list, to keep order
+                            const newNode = new yallist.Node();
+                            newNode.value = foundBufferBlock;
+
+                            currentNode.next = newNode;
+                            newNode.next = nextNode;
+
+                            nextNode.prev = newNode;
+                            newNode.prev = currentNode;
+
+                            newNode.list = this[PRIVATE].allocatedBlocks;
+                            ++this[PRIVATE].allocatedBlocks.length;
+                        }
+                    }
+                }
+
+                // continue searching
+                currentNode = nextNode;
+            }
+        }
+
+        return foundBufferBlock;
+    }
+
+    /**
+     * Attempts to remove an element from the pending allocations,
+     * if possible
+     */
+    maybeAllocateFromPendingAllocations() {
+        if (this[PRIVATE].pendingAllocations.length <= 0) {
+            return;
+        }
+
+        let possibleNextAllocation = this[PRIVATE].pendingAllocations.head;
+        let nextAllocation = null;
+        let bufferMemoryBlock;
+
+        while (possibleNextAllocation !== null &&
+            this[PRIVATE].availablePooledBufferSize > 0) {
+            const allocationRequest = possibleNextAllocation.value;
+
+            if (allocationRequest.requestedSize <= this[PRIVATE].availablePooledBufferSize) {
+                // trying to obtain a buffer block for that allocation
+                bufferMemoryBlock = this.findNextAvailableBufferBlock(allocationRequest.requestedSize);
+                if (bufferMemoryBlock) { // could allocate, can stop going through list
+                    nextAllocation = possibleNextAllocation;
+
+                    const resolvedAllocation = possibleNextAllocation.value;
+
+                    possibleNextAllocation = possibleNextAllocation.next;
+                    this[PRIVATE].pendingAllocations.removeNode(nextAllocation);
+
+                    this[PRIVATE].availablePooledBufferSize -= bufferMemoryBlock.size;
+                    resolvedAllocation.resolve(bufferMemoryBlock);
+                } else {
+                    possibleNextAllocation = possibleNextAllocation.next;
```

Comment:
There are two cases where we iterate to the next allocation block available to evaluate but we aren't asserting that happens in this one scenario.  Can we mock that somehow?

Reply:
Actually, it is tested (allthough indirectly) - it is for cases for which we don't have enough contiguous memory to put a block in. It's tested in some of the other memory allocation tests. The idea here is to go over the list to find a hole in it big enough for the buffer block we have to place.